server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/status/TransportSnapshotsStatusAction.java

```diff
@@ -199,7 +204,20 @@ private void buildResponse(SnapshotsInProgress snapshotsInProgress, SnapshotsSta
                         default:
                             throw new IllegalArgumentException("Unknown snapshot state " + shardEntry.value.state());
                     }
-                    SnapshotIndexShardStatus shardStatus = new SnapshotIndexShardStatus(shardEntry.key, stage);
+                    final SnapshotIndexShardStatus shardStatus;
+                    if (stage == SnapshotIndexShardStage.DONE) {
+                        // Shard snapshot completed successfully so we should be able to load the exact statistics for this
+                        // shard from the repository already.
+                        if (indexIdLookup == null) {
+                            indexIdLookup = entry.indices().stream().collect(Collectors.toMap(IndexId::getName, Function.identity()));
+                        }
+                        final ShardId shardId = shardEntry.key;
+                        shardStatus = new SnapshotIndexShardStatus(shardId, repositoriesService.repository(entry.repository())
+                            .getShardSnapshotStatus(entry.snapshot().getSnapshotId(), indexIdLookup.get(shardId.getIndexName()),
+                                shardId).asCopy());
```

Comment:
why create a copy here?

Reply:
It's just how the constructor `org.elasticsearch.action.admin.cluster.snapshots.status.SnapshotIndexShardStatus#SnapshotIndexShardStatus(org.elasticsearch.index.shard.ShardId, org.elasticsearch.index.snapshots.IndexShardSnapshotStatus.Copy)` works. `Copy` is a different type here it's not an object clone or anything like that. We could definitely look into cleaning this up, there is no need for this complication with the way `IndexShardSnapshotStatus` works these days.